{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# I got the latest greengenes db from the QIIME website \n",
    "# Unzipped it (tar -zxvf file.tar.gz) and put it in the Field folder\n",
    "# This is the assign taxonomy command from QIIME, using the greengenes database at 97%. I am inputting the bacteria\n",
    "# file of OTU sequences.\n",
    "# It is using the default uclust method, at default min 0.9 similarity\n",
    "# It's not working because of an issue making a file in my temporary folder directory.\n",
    "# So, I just made that folder...\n",
    "nprocs = 20\n",
    "!mkdir /tmp/thea_tmp\n",
    "!assign_taxonomy.py -i data/otusn.pick.fasta -t gg_13_8_otus/taxonomy/97_otu_taxonomy.txt \\\n",
    "-r gg_13_8_otus/rep_set/97_otus.fasta -o data/Classification/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[00;38;5;244m\u001b[m\u001b[00;38;5;240motusn.pick_tax_assignments.log\u001b[0m  \u001b[00;38;5;244motusn.pick_tax_assignments.txt\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "ls data/Classification/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU.5687        k__Bacteria; p__Proteobacteria; c__Gammaproteobacteria; o__Legio \bnellales; f__Coxiellaceae; g__; s__     1.00    3\n",
      "OTU.5686        k__Bacteria; p__Planctomycetes; c__Planctomycetia; o__Gemmatales \b; f__Gemmataceae; g__; s__      1.00    3\n",
      "OTU.5689        k__Bacteria; p__OP11; c__; o__; f__; g__; s__   1.00    1\n",
      "OTU.5688        k__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Ricke \bttsiales; f__; g__; s__ 1.00    3\n",
      "OTU.8255        k__Bacteria; p__Gemmatimonadetes; c__Gemm-1; o__; f__; g__; s__  \b1.00    3\n",
      "OTU.1968        k__Bacteria; p__Bacteroidetes; c__Sphingobacteriia; o__Sphingoba \bcteriales; f__; g__; s__        0.67    3\n",
      "OTU.1969        k__Bacteria; p__Armatimonadetes; c__Chthonomonadetes; o__Chthono \bmonadales; f__Chthonomonadaceae; g__; s__       0.67    3\n",
      "OTU.1964        k__Bacteria; p__Elusimicrobia; c__Elusimicrobia; o__Elusimicrobi \bales; f__; g__; s__     1.00    3\n",
      "OTU.1965        k__Bacteria; p__Chloroflexi; c__Ktedonobacteria; o__TK10; f__; g \b__; s__ 1.00    3\n",
      "OTU.1966        k__Bacteria; p__Planctomycetes; c__Phycisphaerae; o__WD2101; f__ \b; g__; s__      1.00    3\n",
      "OTU.1967        k__Bacteria; p__Chlorobi; c__BSV26; o__A89; f__; g__; s__        \b1.00    3\n",
      "OTU.1960        k__Bacteria; p__Acidobacteria; c__Solibacteres; o__Solibacterale \bs; f__; g__; s__        1.00    3\n",
      "\u001b[KOTU.1961        k__Bacteria; p__Proteobacteria; c__Deltaproteobacteria; o__Myxoc \b:\u001b[K"
     ]
    }
   ],
   "source": [
    "!less data/Classification/otusn.pick_tax_assignments.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Joining the classification taxonomy to the biom table\n",
    "! biom add-metadata -i data/otu_table_no211.biom -o data/otu_table_tax_no211.biom \\\n",
    "--observation-metadata-fp data/Classification/otusn.pick_tax_assignments.txt \\\n",
    "--observation-header OTUID,taxonomy,confidence,numrefdbseqs --sc-separated taxonomy "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/opt/anaconda/bin/pyqi\", line 177, in <module>\n",
      "    optparse_main(cmd_obj, argv[1:])\n",
      "  File \"/opt/anaconda/lib/python2.7/site-packages/pyqi/core/interfaces/optparse/__init__.py\", line 276, in optparse_main\n",
      "    result = optparse_cmd(local_argv[1:])\n",
      "  File \"/opt/anaconda/lib/python2.7/site-packages/pyqi/core/interface.py\", line 42, in __call__\n",
      "    return self._output_handler(cmd_result)\n",
      "  File \"/opt/anaconda/lib/python2.7/site-packages/pyqi/core/interfaces/optparse/__init__.py\", line 251, in _output_handler\n",
      "    opt_value)\n",
      "  File \"/opt/anaconda/lib/python2.7/site-packages/pyqi/core/interfaces/optparse/output_handler.py\", line 55, in write_list_of_strings\n",
      "    raise IOError(\"Output path '%s' already exists.\" % option_value)\n",
      "IOError: Output path 'data/otu_table_tax_no211_summary.txt' already exists.\n",
      "Num samples: 119\n",
      "Num observations: 7770\n",
      "Total count: 8330249\n",
      "Table density (fraction of non-zero values): 0.415\n",
      "Table md5 (unzipped): f4eeecfe2048e6a55c2f3ab7fccb336d\n",
      "\n",
      "Counts/sample summary:\n",
      " Min: 8830.0\n",
      " Max: 194356.0\n",
      " Median: 53720.000\n",
      " Mean: 70002.092\n",
      " Std. dev.: 43791.090\n",
      " Sample Metadata Categories: None provided\n",
      " Observation Metadata Categories: taxonomy; confidence; numrefdbseqs\n",
      "\n",
      "Counts/sample detail:\n",
      " S038: 8830.0\n",
      " S242: 13813.0\n",
      " S039: 15038.0\n",
      " S129: 20512.0\n",
      " S245: 21660.0\n",
      " S029: 23329.0\n",
      " S021: 25060.0\n",
      " S239: 25681.0\n",
      " S037: 26554.0\n",
      " S017: 26555.0\n",
      " S014: 26968.0\n",
      " S027: 27587.0\n",
      " S013: 27713.0\n",
      " S018: 28479.0\n",
      " S243: 28671.0\n",
      " S012: 31182.0\n",
      " S131: 31203.0\n",
      " S145: 31203.0\n",
      " S042: 31996.0\n",
      " S246: 32120.0\n",
      " S114: 32298.0\n",
      " S044: 32861.0\n",
      " S016: 34523.0\n",
      " S148: 34617.0\n",
      " S006: 35309.0\n",
      " S247: 35635.0\n",
      " S036: 36097.0\n",
      " S043: 36098.0\n",
      " S023: 36682.0\n",
      " S003: 36728.0\n",
      " S033: 37283.0\n",
      " S034: 37348.0\n",
      " S244: 37578.0\n",
      " S007: 37877.0\n",
      " S233: 39284.0\n",
      " S028: 39352.0\n",
      " S004: 39569.0\n",
      " S022: 39898.0\n",
      " S137: 39955.0\n",
      " S048: 40087.0\n",
      " S241: 40139.0\n",
      " S045: 40986.0\n",
      " S146: 41880.0\n",
      " S117: 42691.0\n",
      " S248: 42810.0\n",
      " S128: 45576.0\n",
      " S041: 46249.0\n",
      " S141: 46890.0\n",
      " S121: 47684.0\n",
      " S011: 47710.0\n",
      " S024: 48814.0\n",
      " S026: 49017.0\n",
      " S002: 49719.0\n",
      " S217: 50293.0\n",
      " S231: 51393.0\n",
      " S008: 51542.0\n",
      " S046: 51708.0\n",
      " S031: 52771.0\n",
      " S102: 53477.0\n",
      " S216: 53720.0\n",
      " S232: 58540.0\n",
      " S116: 58551.0\n",
      " S237: 58807.0\n",
      " S234: 61928.0\n",
      " S138: 63448.0\n",
      " S222: 65700.0\n",
      " S112: 66290.0\n",
      " S209: 66337.0\n",
      " S213: 66819.0\n",
      " S032: 67299.0\n",
      " S109: 67615.0\n",
      " S227: 70308.0\n",
      " S224: 71940.0\n",
      " S103: 73168.0\n",
      " S236: 73434.0\n",
      " S218: 74627.0\n",
      " S113: 74789.0\n",
      " S019: 77203.0\n",
      " S223: 78864.0\n",
      " S229: 79837.0\n",
      " S219: 80119.0\n",
      " S214: 80154.0\n",
      " S124: 81038.0\n",
      " S111: 84124.0\n",
      " S206: 84638.0\n",
      " S001: 86007.0\n",
      " S238: 88192.0\n",
      " S009: 89290.0\n",
      " S107: 89310.0\n",
      " S136: 90128.0\n",
      " S047: 90848.0\n",
      " S147: 90944.0\n",
      " S221: 90967.0\n",
      " S207: 92273.0\n",
      " S108: 96114.0\n",
      " S127: 102935.0\n",
      " S143: 106468.0\n",
      " S119: 107738.0\n",
      " S144: 108677.0\n",
      " S101: 109686.0\n",
      " S104: 114784.0\n",
      " S123: 115443.0\n",
      " S228: 119662.0\n",
      " S118: 126421.0\n",
      " S106: 129553.0\n",
      " S126: 133446.0\n",
      " S208: 139063.0\n",
      " S122: 144272.0\n",
      " S133: 153494.0\n",
      " S139: 154182.0\n",
      " S132: 159371.0\n",
      " S134: 160612.0\n",
      " S142: 168244.0\n",
      " S201: 171706.0\n",
      " S226: 171724.0\n",
      " S202: 177721.0\n",
      " S212: 182268.0\n",
      " S203: 190471.0\n",
      " S204: 194356.0\n"
     ]
    }
   ],
   "source": [
    "! biom summarize-table -i data/otu_table_tax_no211.biom -o data/otu_table_tax_no211_summary.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from biom.parse import parse_biom_table\n",
    "t = parse_biom_table(open(\"data/otu_table_tax_no211.biom\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Trying to add the mapping data too. Oh yeah but there are doubles in the map file.\n",
    "! biom add-metadata -i data/otu_table_tax_no211.biom -o data/otu_table_tax_no211_meta.biom \\\n",
    "--sample-metadata-fp data/MapFile.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "! biom summarize-table -i data/otu_table_tax_no211_meta.biom -o data/otu_table_tax_no211_meta_summary.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num samples: 119\r\n",
      "Num observations: 7770\r\n",
      "Total count: 8330249\r\n",
      "Table density (fraction of non-zero values): 0.415\r\n",
      "Table md5 (unzipped): baeb03877d709768f5f54b8114992354\r\n",
      "\r\n",
      "Counts/sample summary:\r\n",
      " Min: 8830.0\r\n",
      " Max: 194356.0\r\n",
      " Median: 53720.000\r\n",
      " Mean: 70002.092\r\n",
      " Std. dev.: 43791.090\r\n",
      " Sample Metadata Categories: Plants; DayAmdmt; Amdmt; ID; Day\r\n",
      " Observation Metadata Categories: taxonomy; confidence; numrefdbseqs\r\n",
      "\r\n",
      "Counts/sample detail:\r\n",
      " S038: 8830.0\r\n",
      " S242: 13813.0\r\n",
      " S039: 15038.0\r\n",
      " S129: 20512.0\r\n",
      " S245: 21660.0\r\n",
      " S029: 23329.0\r\n",
      " S021: 25060.0\r\n",
      " S239: 25681.0\r\n",
      " S037: 26554.0\r\n",
      " S017: 26555.0\r\n",
      " S014: 26968.0\r\n",
      " S027: 27587.0\r\n",
      " S013: 27713.0\r\n",
      " S018: 28479.0\r\n",
      " S243: 28671.0\r\n",
      " S012: 31182.0\r\n",
      " S131: 31203.0\r\n",
      " S145: 31203.0\r\n",
      " S042: 31996.0\r\n",
      " S246: 32120.0\r\n",
      " S114: 32298.0\r\n",
      " S044: 32861.0\r\n",
      " S016: 34523.0\r\n",
      " S148: 34617.0\r\n",
      " S006: 35309.0\r\n",
      " S247: 35635.0\r\n",
      " S036: 36097.0\r\n",
      " S043: 36098.0\r\n",
      " S023: 36682.0\r\n",
      " S003: 36728.0\r\n",
      " S033: 37283.0\r\n",
      " S034: 37348.0\r\n",
      " S244: 37578.0\r\n",
      " S007: 37877.0\r\n",
      " S233: 39284.0\r\n",
      " S028: 39352.0\r\n",
      " S004: 39569.0\r\n",
      " S022: 39898.0\r\n",
      " S137: 39955.0\r\n",
      " S048: 40087.0\r\n",
      " S241: 40139.0\r\n",
      " S045: 40986.0\r\n",
      " S146: 41880.0\r\n",
      " S117: 42691.0\r\n",
      " S248: 42810.0\r\n",
      " S128: 45576.0\r\n",
      " S041: 46249.0\r\n",
      " S141: 46890.0\r\n",
      " S121: 47684.0\r\n",
      " S011: 47710.0\r\n",
      " S024: 48814.0\r\n",
      " S026: 49017.0\r\n",
      " S002: 49719.0\r\n",
      " S217: 50293.0\r\n",
      " S231: 51393.0\r\n",
      " S008: 51542.0\r\n",
      " S046: 51708.0\r\n",
      " S031: 52771.0\r\n",
      " S102: 53477.0\r\n",
      " S216: 53720.0\r\n",
      " S232: 58540.0\r\n",
      " S116: 58551.0\r\n",
      " S237: 58807.0\r\n",
      " S234: 61928.0\r\n",
      " S138: 63448.0\r\n",
      " S222: 65700.0\r\n",
      " S112: 66290.0\r\n",
      " S209: 66337.0\r\n",
      " S213: 66819.0\r\n",
      " S032: 67299.0\r\n",
      " S109: 67615.0\r\n",
      " S227: 70308.0\r\n",
      " S224: 71940.0\r\n",
      " S103: 73168.0\r\n",
      " S236: 73434.0\r\n",
      " S218: 74627.0\r\n",
      " S113: 74789.0\r\n",
      " S019: 77203.0\r\n",
      " S223: 78864.0\r\n",
      " S229: 79837.0\r\n",
      " S219: 80119.0\r\n",
      " S214: 80154.0\r\n",
      " S124: 81038.0\r\n",
      " S111: 84124.0\r\n",
      " S206: 84638.0\r\n",
      " S001: 86007.0\r\n",
      " S238: 88192.0\r\n",
      " S009: 89290.0\r\n",
      " S107: 89310.0\r\n",
      " S136: 90128.0\r\n",
      " S047: 90848.0\r\n",
      " S147: 90944.0\r\n",
      " S221: 90967.0\r\n",
      " S207: 92273.0\r\n",
      " S108: 96114.0\r\n",
      " S127: 102935.0\r\n",
      " S143: 106468.0\r\n",
      " S119: 107738.0\r\n",
      " S144: 108677.0\r\n",
      " S101: 109686.0\r\n",
      " S104: 114784.0\r\n",
      " S123: 115443.0\r\n",
      " S228: 119662.0\r\n",
      " S118: 126421.0\r\n",
      " S106: 129553.0\r\n",
      " S126: 133446.0\r\n",
      " S208: 139063.0\r\n",
      " S122: 144272.0\r\n",
      " S133: 153494.0\r\n",
      " S139: 154182.0\r\n",
      " S132: 159371.0\r\n",
      " S134: 160612.0\r\n",
      " S142: 168244.0\r\n",
      " S201: 171706.0\r\n",
      " S226: 171724.0\r\n",
      " S202: 177721.0\r\n",
      " S212: 182268.0\r\n",
      " S203: 190471.0\r\n",
      " S204: 194356.0\r\n"
     ]
    }
   ],
   "source": [
    "! cat data/otu_table_tax_no211_meta_summary.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# My final data will be as follows:\n",
    "# My OTU table\n",
    "# This has OTUs and counts per sample\n",
    "[test](data/otu_table.biom)\n",
    "\n",
    "# This file has the sequences of my OTUs\n",
    "> data/otusn.pick.fasta\n",
    "\n",
    "# This file has the assigned taxonomy of my OTUs\n",
    "> data/Classification/otusn.pick_tax_assignments.txt\n",
    "\n",
    "# The mapping file is still\n",
    "> data/MapFile.txt\n",
    "\n",
    "# I have a final .biom file with the taxonomy info in it\n",
    "> data/otu_table_tax.biom\n",
    "> data/otu_table_tax_no211.biom\n",
    "\n",
    "# These should be the only files I care about from now on.\n",
    "# Also note, I have created an aligned file\n",
    "> ssu-aln.bacteria.mask.fasta\n",
    "\n",
    "# And I have also a tree\n",
    "> ssu-aln.bacteria.mask.tre"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
